### PR TITLE
Collection API endpoint

### LINKED ISSUE #137: As a developer, I want an endpoint that returns geo-tagged pollutant data for a single session
**Is your feature request related to a problem? Please describe.**
https://docs.google.com/document/d/1iGCEn3bd37aEJvZQFn6aI0h8mdcPaEzMpTt20h-i83Q/edit#heading=h.p7pm5qhtxvgn

Under the subsection, View session

**Describe the solution you'd like**
Return the pollutant value, lat/lon sufficient for displaying a map

**Testing Outline**
Test that a call returns a list of pollutant values and collection metadata

### PR DESCRIPTION
**Pull request by r-b-g-b**
---
name: Pull Request
about: Request merging changes into the code base
title: ''
labels: ''
reviewers: ''

---

## Checklist
- [x] Add description
- [x] Reference open issue pull request addresses
- [ ] Pass linting check
  - complete on the local machine with `make local.shell` `make quality`
- [ ] Pass functional tests
  - complete on the local machine with `make local.shell` `make test`
- [ ] Request code review
  - Please allow **36 hours** from opening a pull request before merging a pull request- even if it has already recieved an approving review.
- [ ] Address comments on code and resolve requested changes
- [ ] Merge own code

## Description
Issue: Closes #137 Addresses #138 

An endpoint `/collection/#/geo` that returns a pollutant timeseries and metadata for an already stored collection.

### CODE REVIEW

File: api/woeip/apps/core/views.py [hunk 1]
```diff
@@ -2,7 +2,12 @@
...
 
+from rest_framework import generics, viewsets
+from rest_framework.decorators import action
+
 from .constants import Status
+from woeip.apps.core import models, serializers
+
 
 
 def health(_):
```
Comment (nbilenko) [inline api/woeip/apps/core/views.py]: Is this view needed anymore if it's gone from `urls.py`?

File: api/woeip/apps/air_quality/views.py [hunk 2]
```diff
@@ -2,40 +2,99 @@
...
+    serializer_class = serializers.PollutantValueSerializer
+
+
+class SensorViewSet(viewsets.ModelViewSet):
+    queryset = models.Sensor.objects.all()
+    serializer_class = serializers.SensorSerializer
+
+    def get_queryset(self):
+        queryset = models.Sensor.objects.all()
+        sensor_name = self.request.query_params.get("name", None)
```
Comment (nbilenko) [inline api/woeip/apps/air_quality/views.py:88]: Let's be consistent about using `'` vs `"` across the codebase. It looks like there are slightly more `'` so I would lean towards sticking to that for simplicity (and `"""` for docstrings), but no strong preference aside from consistency!

File: api/woeip/apps/air_quality/views.py [hunk 3]
```diff
@@ -2,40 +2,99 @@
...
+
+        return queryset
+
+
+class CollectionViewSet(viewsets.ModelViewSet):
+    queryset = models.Collection.objects.all()
+    serializer_class = serializers.CollectionSerializer
+
+    @action(detail=True, methods=["GET"])
+    def geo(self, request, pk=None):
```
Comment (nbilenko) [inline api/woeip/apps/air_quality/views.py]: I would suggest something like `data` instead of `geo` since this is getting all the data (not just geotags).

File: api/woeip/apps/air_quality/tests/test_api.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+from woeip.apps.air_quality import views
+from woeip.apps.air_quality.tests import factories
+
+request_factory = APIRequestFactory()
+
+
+@pytest.mark.django_db
+def test_collection_list():
+    collection_file = factories.CollectionFileFactory()
```
Comment (nbilenko) [inline api/woeip/apps/air_quality/tests/test_api.py]: Lines 12-22 seem to be unrelated to the collection list test, and nothing happens with `stored_values` after the loop - is this code for another test?

File: api/woeip/apps/air_quality/tests/factories.py [hunk 5]
```diff
@@ -1,14 +1,25 @@
...
+import factory.fuzzy
 import pytz
+import random
 
-from woeip.apps.air_quality.models import Calibration, Collection, CollectionFile, Device, Pollutant, Sensor
+from woeip.apps.air_quality import models
 from woeip.apps.core.tests.factories import UserFactory
 
 
+class FuzzyPoint(factory.fuzzy.BaseFuzzyAttribute):
```
Comment (nbilenko) [inline api/woeip/apps/air_quality/tests/factories.py:13]: This is cool!

File: api/woeip/apps/air_quality/models.py [hunk 6]
```diff
@@ -72,14 +75,21 @@ class CollectionFile(models.Model):
     that captures data from a single sensor. When the file contents are
     processed, the table is updated with the timestamp and processor version.
     """
-    collection = models.ForeignKey(Collection, on_delete=models.CASCADE)
+    collection = models.ForeignKey(
+        Collection,
+        related_name="collection_files",
```
Comment (nbilenko) [inline api/woeip/apps/air_quality/models.py:87]: Nice

**Review by nbilenko — CHANGES_REQUESTED** [review]
Overall looks good to me, just one question about a test and some naming and style suggestions.

**Comment by ifed3** [comment]
@r-b-g-b, this looks amazing, great work! I believe @nbilenko has addressed most of the issues here. The only addition I have refers to issue #142 which I think will address the stylistic differences and also ensure that this PR isn't failing!

File: api/woeip/apps/core/views.py (hunk 1 above)
Comment (r-b-g-b) [reply api/woeip/apps/core/views.py]: Ah good catch, forgot to remove it here.

File: api/woeip/apps/air_quality/views.py (hunk 2 above)
Comment (r-b-g-b) [reply api/woeip/apps/air_quality/views.py:88]: Will be resolved in #143 so I'll leave it for now

File: api/woeip/apps/air_quality/tests/test_api.py (hunk 4 above)
Comment (r-b-g-b) [reply api/woeip/apps/air_quality/tests/test_api.py]: Yeah, this is a bit unfinished. I'd like to add another test that makes sure that the points stored to the db are the same that get returned by the API call. Will work on that now!

File: api/woeip/apps/air_quality/views.py (hunk 3 above)
Comment (r-b-g-b) [reply api/woeip/apps/air_quality/views.py]: Yeah I'm good with that!

File: api/woeip/apps/air_quality/tests/test_api.py (hunk 4 above)
Comment (r-b-g-b) [reply api/woeip/apps/air_quality/tests/test_api.py]: Well the hardest part was finding where this was documented, but looks good now.

**Comment by r-b-g-b** [comment]
Thanks for the input all! I addressed it all and I think it's ready for another look ~~

File: api/woeip/apps/air_quality/tests/test_api.py (hunk 4 above)
Comment (nbilenko) [reply api/woeip/apps/air_quality/tests/test_api.py]: Looks great! I think it might be good to break this up into multiple tests, so that if a test 
fails it's immediately obvious what the problem is. 
Perhaps a good way to organize might be a `TestCollection` class, e.g.

```
class TestCollection(APITestCase):
    @pytest.mark.django_db
    def setUp(self):
        self.collection_file = factories.CollectionFileFactory()
        self.pollutant = factories.PollutantFactory()
        self.n_values = 10
        self.test_values = []
        for _ in range(self.n_values):
            pollutant_value = factories.PollutantValueFactory(
                collection_file=self.collection_file,
                pollutant=self.pollutant,
            )
            self.test_values.append(pollutant_value.value)

    @pytest.mark.django_db
    def test_get_collection_list(self):
        request = request_factory.get("/collection")
        view = views.CollectionViewSet.as_view({"get": "list"})
        response = view(request)
        assert response.status_code == 200

    @pytest.mark.django_db
    def test_get_collection_retrieve(self):
        request = request_factory.get("/collection")
        view = views.CollectionViewSet.as_view(actions={"get": "retrieve"})
        response = view(request, pk=self.collection_file.collection.pk)
        assert response.status_code == 200

    @pytest.mark.django_db
    def test_get_collection_data(self):
        request = request_factory.get("/collection")
        view = views.CollectionViewSet.as_view(actions={"get": "data"})
        response = view(request, pk=self.collection_file.collection.pk)
        assert response.status_code == 200
        stored_values = [pv["value"] for pv in response.data["pollutant_values"]]
        assert stored_values == self.test_values
```

Comment (nbilenko) [reply api/woeip/apps/air_quality/tests/test_api.py]: Also, going to approve the PR since this is a non-blocking improvement suggestion.

**Review by nbilenko — APPROVED** [review]

**Review by ifed3 — APPROVED** [review]
